### ISSUE TITLE
psc-docs: internal error on values without type signatures

### ISSUE THREAD

**Issue by hdgarrood** [purescript/purescript#1834]
(this is a different problem to #1831)

`psc-docs` and `psc-publish` will currently give an internal error if you run them on a module with any exported values which do not have type signatures. For example:

```purescript
module Foo where
foo = 0
```
using `psc-docs` from current master:
```
$ PATH=../purescript/dist/build/psc-docs:$PATH pulp docs
* Generating documentation in /home/harry/code/ps-scratch
psc-docs: An internal error ocurred during compilation: "Docs.Convert.ReExports: lookupValueDeclaration: unexpected result:\nother: []\nident: Ident \"foo\"\ndecls: []"
Please report this at https://github.com/purescript/purescript/issues
* ERROR: Subcommand terminated with exit code 1
```

Previously, it would have just ignored them, so `foo` wouldn't appear in the output. The way I see it, the easiest way forward for now would be to just produce a better error in this case. Perhaps later, we could start using inferred types for exported declarations which are missing documentation; see also https://github.com/purescript/pursuit/issues/21.

On an unrelated note, can we stop calling `show` on the argument to `Language.PureScript.Crash.internalError`? It's already a `String` so I think it would make errors easier to read if we left them as-is.

**Comment by sharkdp** [comment]
I just experienced this error when running `psc-docs` with purescript `master` on an old codebase (it was running fine previously). It took me 15min to figure out what was going on - that is, until I found this bug report. The error message is very irritating.

I know its best-practice to have explicit top-level type declarations, but I guess this bug will be experienced by a lot of people if it doesn't get fixed until 0.8. It basically renders `psc-docs` unusable until you have fixed all your type declarations (or exports) - which could be considered okay, but not with this error message.

Please excuse the non-constructive post.

**Comment by hdgarrood** [comment]
I understand your frustration. On the other hand, there are two points I think should be made:

* whenever you elect to use some software project at master instead of a released version, all bets are off, effectively. If the maintainers thought it was ready to go, it would have been released.
* the error does state pretty clearly that it's an internal error. Not understanding what that means is perhaps reasonable, though. What "internal error" basically means is that, at the time the code was written, the scenario that led to this error being thrown did not occur, or was considered impossible, to whoever was writing this code (in this case, me). Given that, at the time I wrote the code, I did think this error would never be thrown, I actually think it's very appropriate. Of course, with hindsight, it's rather bad, but that's beside the point.

If the problem is that the meaning of "internal error" is not clear enough, then maybe we should address that.

It's also worth noting that the current behaviour is arguably an improvement over the previous behaviour, which was to silently discard declarations without type signatures. So it wasn't really working perfectly before at all.

**Comment by sharkdp** [comment]
I'm sorry if I sounded like being frustrated. 

> whenever you elect to use some software project at master instead of a released version, all bets are off, effectively. If the maintainers thought it was ready to go, it would have been released.

Absolutely. My comment was solely based on the "0.9.0-milestone" label and my only intent was to raise awareness that this might be experienced by a lot of people in case it would be released as 0.8.

> It's also worth noting that the current behaviour is arguably an improvement over the previous behaviour, which was to silently discard declarations without type signatures. So it wasn't really working perfectly before at all.

I agree. Failing with an error is better than to silently discard. In my opinion, a warning would be even better (`Warning in module Foo at ..: Ignoring exported function "foo" without explicit type annotation.`).

I would prefer the warning because the error basically means that `psc-docs` fails on code which compiles (admittedly though, with a compiler warning).

**Comment by garyb** [comment]
Is there a difficulty with using the inferred type when there is no declaration? Without even glancing at the code it seems like it should be relatively straightforward, assuming the make stage is invoked before docs are generated?

**Comment by garyb** [comment]
Ok, so, it doesn't make, that would indeed be a problem :smile:

**Comment by hdgarrood** [comment]
Do you think it should change to use make if we were going to start inferring types?

**Comment by garyb** [comment]
I think it has to really, `psc-publish` does so, doesn't it?

**Comment by hdgarrood** [comment]
Nope, `psc-publish` uses `Language.PureScript.Docs.ParseAndDesugar` as well

**Comment by garyb** [comment]
Ah right. I guess the same problem applies there too then?

**Comment by hdgarrood** [comment]
Yeah, previously, stuff without explicit type signatures would just be silently omitted, and in current master, it causes this internal error right now.

**Comment by garyb** [comment]
Integrating a make step shouldn't be too hard, I'm just not sure what the best way of handling it is - it seems like it might be weird for `psc-{publish,docs}` to do much of the same work as `psc`.

We'd also need to copy a bunch of command line arguments from `psc` unless we did something so that no JS files are produced, only the externs. That way all FFI-related and optimisation flags can be omitted, we'd just need `--output`. It also shouldn't interfere with `psc` then either, as if the js files are missing from the output directory `psc` will rebuild as it should.

**Comment by garyb** [comment]
@paf31 we talked a bit more about this on IRC afterwards. Another suggestion was to just do the compilation process in memory. It would be slow, as there'd be no reading of externs, etc. but also would keep things simpler for `psc-{publish,docs}`.

The slowness wouldn't be a problem when generating docs/pursuit data normally, as you don't do that as part of the development process, however it would be a problem for etags/ctags production.

So then an option would be to omit inferred types from the tags files, but then that gives us another route to deal with in `docs`, and we still need to prevent the internal errors from being emitted.

So... any thoughts?

**Comment by hdgarrood** [comment]
There's no problem for etags/ctags, they use a different code path, so they wouldn't be affected. The tags files have nothing to do with types anyway, right? They're just for finding where functions are defined, I think.

After typechecking, I was imagining that we a) infer missing types and add them to the AST somehow, and b) modify `Language.PureScript.Docs.Convert` so that it picks up these inferred types. Does that sound doable?

**Comment by garyb** [comment]
Oh! Okay, that's cool then. In my ignorance I assumed they included type info too.

Definitely, every declaration will be wrapped in a `TypedValue` after typechecking, so you can get the information from there.

**Comment by hdgarrood** [comment]
Oh also, @sharkdp, I think the 0.9.0 milestone really means "at the latest 0.9.0, but ship it in the first release that we can once it's done". If it were a breaking change, that would actually be 0.9.0. Since it's just a bugfix, it will probably be shipped pretty soon after 0.8 (if not in 0.8 itself).

**Comment by sharkdp** [comment]
I didn't know that. Thanks for the clarification.
